core/clipboard.ts

```diff
@@ -68,23 +108,35 @@ export function paste(): ICopyable | null {
  *     duplication failed.
  * @internal
  */
-export function duplicate(toDuplicate: ICopyable): ICopyable | null {
+export function duplicate<
+  U extends ICopyData,
+  T extends ICopyable<U> & IHasWorkspace,
+>(toDuplicate: T): T | null {
   return TEST_ONLY.duplicateInternal(toDuplicate);
 }
 
 /**
  * Private version of duplicate for stubbing in tests.
  */
-function duplicateInternal(toDuplicate: ICopyable): ICopyable | null {
-  const oldCopyData = copyData;
-  copy(toDuplicate);
-  if (!copyData || !source) return null;
-  const pastedThing =
-    globalRegistry
-      .getObject(globalRegistry.Type.PASTER, copyData.paster, false)
-      ?.paste(copyData, source) ?? null;
-  copyData = oldCopyData;
-  return pastedThing;
+function duplicateInternal<
+  U extends ICopyData,
+  T extends ICopyable<U> & IHasWorkspace,
+>(toDuplicate: T): T | null {
+  const oldCopyData = stashedCopyData;
+  const oldWorkspace = stashedWorkspace;
+
+  const data = copy(toDuplicate);
+
+  // I hate side effects.
+  stashedCopyData = oldCopyData;
+  stashedWorkspace = oldWorkspace;
+
+  if (!data) return null;
+  return paste(data, toDuplicate.workspace) as T;
+}
+
+interface IHasWorkspace {
```

Comment:
I'm kind of confused by this. Shouldn't the duplicate method take in a `workspace` parameter like `paste` does? This method is already `@internal` so you could just update the callers to pass the workspace and not even worry about adding this particular bit of backwards compatibility.

Reply:
Can I just delete `duplicate` and `copy` since they're internal? I originally proposed maing them backwards compatible but deprecating them because I had forgotten they were internal by the time I wrote my design doc :P

But honestly I think it's better to handle this logic within the shortcuts/context menu options.

Reply:
Discussed offline. These are `@internal` but don't have underscores + are useful enough we think people are probably using them. We're going to deprecate v11, and remove v12.